design/blueprint_store.md

```diff
@@ -1,50 +1,227 @@
+### Blueprint APIs
 
-The blueprint store is organized flatly by just blueprint id, e.g. `viewport.children = […]` where `children` is a *ui component (or “blueprint component”?). In this case `children` contain other blueprint ids (the views in the viewport).
+The blueprint APIs will follow a declarative structure, allowing users to build up a hierarchical blueprint object
+that matches the Rerun concept map:
+```yaml
+App:
+    Viewport:
+        Container:
+            SpaceView:
+                DataGroup:
+                    Data
+                    Data
+                    ...
+            Container:
+                SpaceView:
+                    DataGroup:
+                        Data
+                        Data
+                        ...
+                SpaceView:
+                    DataGroup:
+                        Data
+                        Data
+                        ...
+            ...
+```
 
-The blueprint store has exactly one timeline: `ui_time` which is the local time of the application. This can then be used for undo and redo.
+A theoretical python API might look like:
+```python
+blueprint = rrb.App(
+    expand_panels=False,
+    time_control=rrb.TimeControl(
+        timeline="sim_time",
+        play_state="paused",
+    ),
+    viewport=rrb.VerticalLayout(
+        content=[
+            rrb.View3D(
+                root="world",
+                content=[
+                    rrb.Data("./points", visible_history=10).override(radius=0.1, color="blue"),
+                    rrb.Data("./camera").override(image_plane=3),
+                ],
+            ),
+            rrb.HorizontalLayout(
+                content=[
+                    rrb.View2D(
+                        root="world/camera/image",
+                        content=rrb.DataGroup(
+                            rrb.Data("."),
+                            rrb.Points2D("world/points"),
+                        ).default(radius=0.2),
+                    ),
+                    rrb.ViewTimeSeries("metrics", content=rrb.RecursiveData(".")),
+                ],
+            ),
+        ],
+    ),
+)
 
-## Viewer
-The viewer state would be completely driven by the blueprint store and the data store. Each frame we would basically query the blueprint store about the current state of the ui, which would then dictate the queries done to the data store.
+rr.spawn(blueprint)
+```
+
+The assorted objects used in blueprint construction are:
+ - `App`: Container for top-level application state such as panel-visibility, menus, etc.
+ - `TimeControl`: Specific state relevant to the time controls.
+ - `Tile`: Common base-class between `rrb.Container` and `rrb.SpaceView` (follows egui-tile-tree)
+    - `Container`: A layout of components (interface only)
+        - `HorizontalLayout`
+        - `VerticalLayout`
+        - ... additional layouts
+    - `SpaceView`: An actual view of data
```

Comment:
```suggestion
    - `SpaceView`: An actual view of data in a coordinate space
```

Reply:
is a table view of raw data a coordinate space? I think that stretches it a bit too far

Reply:
It does stretch it a bit far but I also think that is what the "Space" in SpaceView refers to. So that might be an argument to not call it space view.

Reply:
Yeah, coordinate-space isn't quite right, but there's some generalized analogy to coordinate-space that probably is, and in the mean-time it's a reasonable place-holder.  I think a Space is more loosely "a set of data that can all be transformed/mapped into a single *self-consistent* representation.